### ISSUE TITLE
[Upgrade from mixed-storage LXD] LXD won't restart after upgrade to 2.10.1

### ISSUE THREAD

**Issue by sbworth** [lxc/lxd#3026]
Hello,

I believe that I have a variant of the problem seen in issue #3024 which I have been following with interest. After upgrade to 2.10.1 from 2.8.x, lxd cannot start up.

# Required information

 * Distribution: Ubuntu
 * Distribution version: 16.04LTS
 * The output of "lxc info" or if that fails:
   * Kernel version:
   * LXC version: 2.10.1
   * LXD version: 2.10.1
   * Storage backend in use: LVM (thinpools)

# Issue description

I have two systems, sys1 and sys2. Sys1 is using dir storage, while sys2 is using LVM. 

With sys1, I migrated from 2.8.x to 2.9.x and then to 2.10.x. After resolving an issue with a change in profile inheritance of the disk device after the 2.9.x upgrade, sys1 seems to have upgraded to 2.10.x ok.

With sys2, I migrated directly from 2.8.x to 2.10.x. This was inadvertent, as I had just sorted out the 2.9.x issue on sys1 and intended to move sys2 to 2.9.x. When lxd attempted to restart, the lxc command line client stopped responding.

Checking /var/log/lxd/lxd.log, we see:
```
lvl=info msg="LXD 2.10.1 is starting in normal mode" path=/var/lib/lxd t=2017-03-06T14:34:02-0500
lvl=warn msg="CGroup memory swap accounting is disabled, swap limits will be ignored." t=2017-03-06T14:34:02-0500
lvl=info msg="Kernel uid/gid map:" t=2017-03-06T14:34:02-0500
lvl=info msg=" - u 0 0 4294967295" t=2017-03-06T14:34:02-0500
lvl=info msg=" - g 0 0 4294967295" t=2017-03-06T14:34:02-0500
lvl=info msg="Configured LXD uid/gid map:" t=2017-03-06T14:34:02-0500
lvl=info msg=" - u 0 100000 65536" t=2017-03-06T14:34:02-0500
lvl=info msg=" - g 0 100000 65536" t=2017-03-06T14:34:02-0500
lvl=warn msg="Database already contains a valid entry for the storage pool: lxd." t=2017-03-06T14:34:03-0500
lvl=warn msg="Storage volumes database already contains an entry for the container." t=2017-03-06T14:34:03-0500
lvl=info msg="LXD 2.10.1 is starting in normal mode" path=/var/lib/lxd t=2017-03-06T14:44:02-0500
lvl=warn msg="CGroup memory swap accounting is disabled, swap limits will be ignored." t=2017-03-06T14:44:02-0500
lvl=info msg="Kernel uid/gid map:" t=2017-03-06T14:44:02-0500
lvl=info msg=" - u 0 0 4294967295" t=2017-03-06T14:44:02-0500
lvl=info msg=" - g 0 0 4294967295" t=2017-03-06T14:44:02-0500
lvl=info msg="Configured LXD uid/gid map:" t=2017-03-06T14:44:02-0500
lvl=info msg=" - u 0 100000 65536" t=2017-03-06T14:44:02-0500
lvl=info msg=" - g 0 100000 65536" t=2017-03-06T14:44:02-0500
lvl=warn msg="Database already contains a valid entry for the storage pool: lxd." t=2017-03-06T14:44:03-0500
lvl=warn msg="Storage volumes database already contains an entry for the container." t=2017-03-06T14:44:03-0500
```
# journalctl -u lxd

```
Mar 06 14:34:02 sys2 systemd[1]: Starting LXD - main daemon...
Mar 06 14:34:02 sys2 lxd[4416]: lvl=warn msg="CGroup memory swap accounting is disabled, swap limits will be ignored." t=2017-03-06T14:34:02-0500
Mar 06 14:34:03 sys2 lxd[4416]: lvl=warn msg="Database already contains a valid entry for the storage pool: lxd." t=2017-03-06T14:34:03-0500
Mar 06 14:34:03 sys2 lxd[4416]: lvl=warn msg="Storage volumes database already contains an entry for the container." t=2017-03-06T14:34:03-0500
Mar 06 14:34:13 sys2 lxd[4416]: error: device or resource busy
Mar 06 14:34:13 sys2 systemd[1]: lxd.service: Main process exited, code=exited, status=1/FAILURE
Mar 06 14:44:02 sys2 lxd[4417]: error: LXD still not running after 600s timeout.
Mar 06 14:44:02 sys2 systemd[1]: lxd.service: Control process exited, code=exited status=1
Mar 06 14:44:02 sys2 systemd[1]: Failed to start LXD - main daemon.
Mar 06 14:44:02 sys2 systemd[1]: lxd.service: Unit entered failed state.
Mar 06 14:44:02 sys2 systemd[1]: lxd.service: Failed with result 'exit-code'.
Mar 06 14:44:02 sys2 systemd[1]: lxd.service: Service hold-off time over, scheduling restart.
Mar 06 14:44:02 sys2 systemd[1]: Stopped LXD - main daemon.
Mar 06 14:44:02 sys2 systemd[1]: Starting LXD - main daemon...
Mar 06 14:44:02 sys2 lxd[8637]: lvl=warn msg="CGroup memory swap accounting is disabled, swap limits will be ignored." t=2017-03-06T14:44:02-0500
Mar 06 14:44:03 sys2 lxd[8637]: lvl=warn msg="Database already contains a valid entry for the storage pool: lxd." t=2017-03-06T14:44:03-0500
Mar 06 14:44:03 sys2 lxd[8637]: lvl=warn msg="Storage volumes database already contains an entry for the container." t=2017-03-06T14:44:03-0500
Mar 06 14:44:13 sys2 lxd[8637]: error: device or resource busy
Mar 06 14:44:13 sys2 systemd[1]: lxd.service: Main process exited, code=exited, status=1/FAILURE
```
Sample of `/var/lib/lxd/containers`:
```
drwx------+ 5 root   root    4096 Jan  2 15:59 astro3
lrwxrwxrwx  1 root   root      16 Jan  2 11:35 astro3.lv -> /dev/lxd/astro3
drwxr-xr-x+ 5 root   root    4096 Jan  9 12:13 vault
lrwxrwxrwx  1 root   root      14 Jan 11 15:01 vault.lv -> /dev/lxd/vault
drwx------  2 root   root    4096 Feb 15 13:48 vgate1
lrwxrwxrwx  1 root   root      15 Feb 15 13:48 vgate1.lv -> /dev/lxd/vgate1
drwxr-xr-x+ 5 root   root    4096 Jan 18 13:52 vpn1
lrwxrwxrwx  1 root   root      13 Jan 25 16:22 vpn1.lv -> /dev/lxd/vpn1
```
File tree listing of `/var/lib/lxd/storage-pools`:
```
.
└── lxd
    └── containers
```
That is, the storage-pools area is empty. (Were the container rootfs links supposed to be migrated to `storage-pools`?)

The images area seems untouched:
```
root@sol2:/var/lib/lxd/containers# ls /var/lib/lxd/images/
11fc1b1d39b9f9cd7e9491871f1421ac4278e1d599ecf5d180f2a6e2483bd172
11fc1b1d39b9f9cd7e9491871f1421ac4278e1d599ecf5d180f2a6e2483bd172.lv
11fc1b1d39b9f9cd7e9491871f1421ac4278e1d599ecf5d180f2a6e2483bd172.rootfs
18e7ed74d0d653894f65343afbc35b92c6781933c273943d882c36a5c5535533
18e7ed74d0d653894f65343afbc35b92c6781933c273943d882c36a5c5535533.lv
457a80ea4720900b69e5542cea5351f58021331bc96e773e4855a3e2ce1e6595
457a80ea4720900b69e5542cea5351f58021331bc96e773e4855a3e2ce1e6595.lv
457a80ea4720900b69e5542cea5351f58021331bc96e773e4855a3e2ce1e6595.rootfs
543e662b70958f5b87f68b20eb0a205d8c4b14c41f80699e9a98b3b851883d15
543e662b70958f5b87f68b20eb0a205d8c4b14c41f80699e9a98b3b851883d15.lv
543e662b70958f5b87f68b20eb0a205d8c4b14c41f80699e9a98b3b851883d15.rootfs
a570ce23e1dae791e7b8b2f2bcb98c1404273e97c7a1fb972bf0f5835ac3e869
a570ce23e1dae791e7b8b2f2bcb98c1404273e97c7a1fb972bf0f5835ac3e869.lv
b5b03165de7c450f5f9793c8b2eb4a364fbd81124a01511f854dd379eef52abb
b5b03165de7c450f5f9793c8b2eb4a364fbd81124a01511f854dd379eef52abb.rootfs
bfd17410a8c7fe6397dba3e353a23001243bc43af87acf25544d6b0ab624f9f8
bfd17410a8c7fe6397dba3e353a23001243bc43af87acf25544d6b0ab624f9f8.rootfs
d7c16c4fedd3308b5bffdb91f491b8458610c6115d37ace8ba4bcf5c29b23cc6
d7c16c4fedd3308b5bffdb91f491b8458610c6115d37ace8ba4bcf5c29b23cc6.lv
d7c16c4fedd3308b5bffdb91f491b8458610c6115d37ace8ba4bcf5c29b23cc6.rootfs
e12c3c1aed259ce62b4a5e8dc5fe8b92d14d36e611b3beae3f55c94df069eeed
e12c3c1aed259ce62b4a5e8dc5fe8b92d14d36e611b3beae3f55c94df069eeed.lv
ff52f536d2896f358bc913d592828ecf1b39fae45e4ee4825930091e8793ac28
ff52f536d2896f358bc913d592828ecf1b39fae45e4ee4825930091e8793ac28.rootfs
```

Output from `pvs` and `vgs` and -- highly edited for readability -- output from `lvs`:
```
  PV         VG   Fmt  Attr PSize   PFree  
  /dev/sda5  dat1 lvm2 a--  931.13g 181.13g
  /dev/sda6  lxd  lvm2 a--    2.56t      0 
```
```
  VG   #PV #LV #SN Attr   VSize   VFree  
  dat1   1   1   0 wz--n- 931.13g 181.13g
  lxd    1  42   0 wz--n-   2.56t      0 
```
```
 LV        VG   Attr       LSize   Pool   Origin   Data%  Meta%
 LXDPool  lxd  twi-aotz--   2.56t                   3.91   2.12
 astro3   lxd  Vwi-aotz--  10.00g LXDPool          20.69 
 vault    lxd  Vwi-aotz--  10.00g LXDPool          12.34
 vgate1   lxd  Vwi-a-tz-- 300.00g LXDPool           1.85
 vpn1     lxd  Vwi-aotz-- 300.00g LXDPool           1.88
```

Data from lxd.db:
```
sqlite> select * from storage_pools;
1|lxd|lvm
sqlite> select * from storage_pools_config;
166|1|volume.size|300GB
167|1|size|21GB
168|1|source|lxd
169|1|lvm.thinpool_name|LXDPool
170|1|lvm.vg_name|lxd
sqlite> select * from storage_volumes;
1|astro3|1|0
sqlite> select * from storage_volumes_config;
67|1|block.filesystem|ext4
68|1|size|300GB
```
It looks somewhat odd to me that host astro3 has an entry in the storage_volumes tables when nothing else does. It does differ in being a privileged container.

Any help you can provide to get regular access restored will be greatly appreciated. For the moment, the containers continue to provide their services. Let me know if I can provide any other useful data or perform any non-destructive tests.

**Comment by stgraber** [comment]
The error in your case appears to be "error: device or resource busy".
I don't know if @brauner has an idea of what may be causing this on LVM and if we can work around that issue too.

The reason why you have that container listed but not the others is simply because it's the first in alphabetical order. It may be that container which failed to get converted or the next one, I'm not sure.

@brauner did you do LVM upgrade tests from 2.8 to 2.10.1 with running containers? I wonder if it's just an issue of a mountpoint needing some convincing.

**Comment by brauner** [comment]
I'm pretty sure that this might again be solved by using lazy umounts.

**Comment by stgraber** [comment]
There is a very good chance that rebooting your system will fix the issue as all containers would then get stopped, all mounts undone and so give LXD a clean environment to upgrade on reboot. But if you're willing to wait a bit we can try to fix this for running containers and give you an updated lxd binary to try.

**Comment by brauner** [comment]
And yes, I've run upgrade tests from 2.8. to 2.10.1 with LVM: https://asciinema.org/a/104652

**Comment by stgraber** [comment]
@sbworth 

https://dl.stgraber.org/lxd-3026
sha256: 036428ed214b786c771cfa64dabf888be492a44ec3763607c9656bdc46d850f0

To run:
```
systemctl stop lxd.service lxd.socket
./lxd --debug --group lxd
```

**Comment by sbworth** [comment]
Will try as soon as I can get to a console.

**Comment by sbworth** [comment]
Is lxd-3026 intended to fix the problem or just provide diagnostics?

**Comment by stgraber** [comment]
It should fix the problem, or at least get past the error we saw earlier.

What's the output?

**Comment by brauner** [comment]
I suspect that the point of failure is that the umount didn't succeed because something is holding the mountpoint busy. If that's the case @stgraber's binary should fix this problem.

**Comment by sbworth** [comment]
Sorry, but I will have to put off testing until I have the chance to warn some users, which will likely mean putting off the actual test until tomorrow morning (EST). Thank you for your very rapid response.

**Comment by stgraber** [comment]
@sbworth Note that the upgrade will not stop any running containers, so assuming that your LXD daemon is currently offline, it really can't make things any worse.

**Comment by sbworth** [comment]
OK. Will try momentarily.

**Comment by sbworth** [comment]
```
# ./lxd --debug --group lxd
INFO[03-06|16:48:08] LXD 2.10.1 is starting in normal mode    path=/var/lib/lxd
WARN[03-06|16:48:08] CGroup memory swap accounting is disabled, swap limits will be ignored. 
INFO[03-06|16:48:08] Kernel uid/gid map: 
INFO[03-06|16:48:08]  - u 0 0 4294967295 
INFO[03-06|16:48:08]  - g 0 0 4294967295 
INFO[03-06|16:48:08] Configured LXD uid/gid map: 
INFO[03-06|16:48:08]  - u 0 100000 65536 
INFO[03-06|16:48:08]  - g 0 100000 65536 
DBUG[03-06|16:48:08] Initializing and checking storage pool "lxd". 
DBUG[03-06|16:48:08] Initializing an LVM driver. 
DBUG[03-06|16:48:08] Checking LVM storage pool "lxd". 
DBUG[03-06|16:48:08] Checked LVM storage pool "lxd". 
DBUG[03-06|16:48:08] Applying patch: storage_api 
WARN[03-06|16:48:08] Database already contains a valid entry for the storage pool: lxd. 
WARN[03-06|16:48:08] Storage volumes database already contains an entry for the container. 
error: exit status 5
```

**Comment by brauner** [comment]
Exit status `5` means that either the physical volume, volume group or logical volume is not found.

**Comment by sbworth** [comment]
`/var/lib/lxd/storage-pools/lxd/containers` now contains:
```
drwx------+ 5 root   root   4096 Jan  2 15:59 apollo3
drwxr-xr-x+ 5 100000 100000 4096 Jan  3 13:28 apps4
drwxr-xr-x+ 4 100000 100000 4096 Nov 24 11:33 apps4-old
```

**Comment by sbworth** [comment]
I have no container apps4-old; so I am rather confused. Investigating contents.
```
.
├── apollo3
│   ├── backup.yaml
│   ├── lost+found
│   ├── metadata.yaml
│   ├── rootfs
│   └── templates
├── apps4
│   ├── backup.yaml
│   ├── lost+found
│   ├── metadata.yaml
│   ├── rootfs
│   └── templates
└── apps4-old
    ├── metadata.yaml
    ├── rootfs
    └── templates
```
Could it be picking up old data from a previous migration to LVM from dir storage?

**Comment by stgraber** [comment]
Hmm, that's confusing, I wonder if you have that "apps4-old" container in the database but it wasn't showing up due to broken storage.

What does this give you:

    sqlite3 /var/lib/lxd/lxd.db "SELECT * FROM containers;"

(You'll need to install the "sqlite3" package if you don't already have it installed)

**Comment by sbworth** [comment]
First line in the table:
```
3|apps4-old|2|0|0|0|1480005211|2016-11-30 14:52:15.442158778+00:00
```
No evidence of it in `/var/lib/lxd/containers`.

Looking for other strays.

**Comment by stgraber** [comment]
Ok, so that explains the migration problem. Are you seeing any other such containers in the database list?

**Comment by stgraber** [comment]
If not, the following should do the trick to get rid of it from the database, unblocking the migration:
```
sqlite3 /var/lib/lxd/lxd.db
  PRAGMA foreign_keys=ON;
  DELETE FROM containers WHERE name="apps4-old";

  <ctrl+d> to exit the interactive DB session
```

**Comment by sbworth** [comment]
Remove apps4-old and another stray, both dating back to days after original installation and probably before migration to LVM backing store. 

Ran your lxd again and it died in the same way. 
```
DBUG[03-06|17:12:39] Checking LVM storage pool "lxd". 
DBUG[03-06|17:12:39] Checked LVM storage pool "lxd". 
DBUG[03-06|17:12:39] Applying patch: storage_api 
WARN[03-06|17:12:39] Database already contains a valid entry for the storage pool: lxd. 
WARN[03-06|17:12:39] Storage volumes database already contains an entry for the container. 
error: exit status 5
```

Looking for more strays.

Some of my containers mount data from other LVM volumes not managed by LXD. Could that be a problem?

**Comment by stgraber** [comment]
No, container devices don't impact the migration process.

**Comment by sbworth** [comment]
Does the migration occur in alphabetical order?

**Comment by stgraber** [comment]
Yeah, I believe the containers list we fetch from the database is sorted in alphabetical order.

**Comment by stgraber** [comment]
I'm looking at the code now to see where else we're spawning a subprocess without better logging things...

**Comment by brauner** [comment]
This could be `lvrename` failing because the container has already been succesfully moved.
If that's the case I can send a patch.

**Comment by stgraber** [comment]
@sbworth What does "lvs" show you? I'd expect the containers that show up in /var/lib/lxd/storage-pools/containers to have a "containers" prefix to their name while the others don't. Is your first container renamed and not the others?

**Comment by sbworth** [comment]
What are we `lvrename`ing? 
The two successfully moved containers are now:
```
lrwxrwxrwx 1 root root 8 Mar  6 16:48 containers_apollo3 -> ../dm-18
lrwxrwxrwx 1 root root 7 Mar  6 16:48 containers_apps4 -> ../dm-8
```
The others are all untouched.

**Comment by brauner** [comment]
@sbworth, @stgraber means the output of `lvs` on your system so for you LXD vg: `sudo lvs lxd`.

**Comment by stgraber** [comment]
Ok, so two of them did get renamed, which is causing the issue. Ok, that makes sense. @brauner is working on a fix for that, I'll give you a new binary build once I have it.

**Comment by sbworth** [comment]
OK. That is just another way of viewing what I showed you:
```
LV                  VG   Attr        LSize   Pool   Origin   Data%  Meta%
 containers_apollo3  lxd  Vwi-aotz--  10.00g LXDPool  20.69
 containers_apps4    lxd  Vwi-aotz--  10.00g LXDPool  53.16
```
All the other still have no containers prefix.

**Comment by brauner** [comment]
I just wanted to make sure that we're seeing the right volume group.

**Comment by sbworth** [comment]
I have to catch a train soon; so I will be out of contact for at least 80 minutes. (What TZ are you guys in?)
Thanks again for your efforts.

**Comment by stgraber** [comment]
@brauner is in Germany, I'm in Canada (east coast).

**Comment by brauner** [comment]
https://github.com/lxc/lxd/pull/3027

**Comment by stgraber** [comment]
https://dl.stgraber.org/lxd-3026
sha256: ab1874f7f33761196969a80c398c2ce542b271499179973770624e038738e37f

**Comment by stgraber** [comment]
Updated version for that last patch.

https://dl.stgraber.org/lxd-3026
sha256: e22e69ea1dae2f113677b89e13d8f571367ec5fde90675e3b3e00c239d7ec44b

**Comment by stgraber** [comment]
And another one to include a few extra checks that @brauner added.

https://dl.stgraber.org/lxd-3026
sha256: 41c7c751ce952989144224ffcc855b0f1883ef123727f35518c3620c02fd04af

@sbworth let us know when you're back.
We're hoping we can include this fix in LXD 2.11 (to be released tomorrow).

**Comment by stgraber** [comment]
And again with improved cleanup on failure.

https://dl.stgraber.org/lxd-3026
sha256: d8b098bc0177a8ab10b6c4e86b1fbbec93b43983267f9f314f3153838f7edab7

**Comment by sbworth** [comment]
Latest trial gets further, but still dies with exit 5. Migrated 4 more containers before dying, but still many more to go. Will continue looking for obvious problem sources.

(Sorry for getting back so late. I have a one year old at home. Plans evolve rapidly.)

**Comment by brauner** [comment]
@sbworth, if possible it would be great if you could provide the new output after having run the latest binary that @stgraber uploaded. This would involve:

- the on-disk layout of the volume group and its logical volumes
- the db entries for storage_pools, storage_pools_config, storage_volumes, storage_volumes_config
- the LXD on-disk layout:
  - ls -al /var/lib/lxd/{containers,snapshots,images}
  - ls -al /var/lib/lxd/storage-pools and subdirectories

What we'd need now is to find out why there is still an exit status 5.

**Comment by stgraber** [comment]
I merged @brauner's fixes since those are good to have anyway. I'm preparing a commit which will convert our codebase to actually log the command on failure, a binary with that applied should make failure a bit more understandable.

**Comment by sbworth** [comment]
Sorry if I was not clear, but I ran the latest stgraber patched lxd. That is what gave me 4 more containers before bailing. I will send along the requested data as soon as I pull it together.

**Comment by sbworth** [comment]
This is going to take some time. I am fine with posting the bits that seem relevant, but I am very reluctant to post the entire dataset to a public forum without anonymization and I don't have a script ready to do that. Writing such a script now. Will also enable me to quickly generate summary data for any follow-up tests as well.

**Comment by brauner** [comment]
@sbworth, oh sure. Don't feel pressured to provide this information directly here. If you'd rather provide it in private (anonymized or non-anonymized) to @stgraber and me you can just send a private mail.

**Comment by stgraber** [comment]
I have a LXD build which should add much of the command logging now, lets see if that tells us what's going on.

**Comment by stgraber** [comment]
@sbworth 
https://dl.stgraber.org/lxd-3026
sha256: 648d9999383cd87e5412204b3c0962ddec0984a752d2995f27db06ca4c215c5a

**Comment by sbworth** [comment]
Is it expected that LXD will migrate the images folder as well? If I am interpreting this correctly, I think I have old pre-LVM images living side-by-side with the symlinks to the logical volumes that should currently represent them.
```
-rw-r--r-- 1 root root       844 Mar  3 21:37 11fc1b1d39b9f9cd7e9491871f1421ac4278e1d599ecf5d180f2a6e2483bd172
lrwxrwxrwx 1 root root        73 Mar  3 21:41 11fc1b1d39b9f9cd7e9491871f1421ac4278e1d599ecf5d180f2a6e2483bd172.lv -> /dev/lxd/11fc1b1d39b9f9cd7e9491871f1421ac4278e1d599ecf5d180f2a6e2483bd172
-rw-r--r-- 1 root root 153494344 Mar  3 21:40 11fc1b1d39b9f9cd7e9491871f1421ac4278e1d599ecf5d180f2a6e2483bd172.rootfs
-rw-r--r-- 1 root root 199032954 Jan 10 10:45 18e7ed74d0d653894f65343afbc35b92c6781933c273943d882c36a5c5535533
lrwxrwxrwx 1 root root        73 Jan 10 10:45 18e7ed74d0d653894f65343afbc35b92c6781933c273943d882c36a5c5535533.lv -> /dev/lxd/18e7ed74d0d653894f65343afbc35b92c6781933c273943d882c36a5c5535533
-rw-r--r-- 1 root root       600 Jan  9 13:47 457a80ea4720900b69e5542cea5351f58021331bc96e773e4855a3e2ce1e6595
lrwxrwxrwx 1 root root        73 Jan  9 13:47 457a80ea4720900b69e5542cea5351f58021331bc96e773e4855a3e2ce1e6595.lv -> /dev/lxd/457a80ea4720900b69e5542cea5351f58021331bc96e773e4855a3e2ce1e6595
-rw-r--r-- 1 root root  98614308 Jan  9 13:47 457a80ea4720900b69e5542cea5351f58021331bc96e773e4855a3e2ce1e6595.rootfs
```

**Comment by sbworth** [comment]
Sorry, my interpretation must be wrong. The time stamps all indicate recent access. Apologies for publically arguing with myself.

**Comment by stgraber** [comment]
Yeah, LXD will migrate images too, but it does that after it's done with containers and snapshots.

When it does, /var/lib/lxd should only contain the image files themselves and the mountpoint for the LVs will be created under /var/lib/lxd/storage-pools/images/ and those symlinks should then disappear.

**Comment by sbworth** [comment]
OK. Thanks for the clarification.
I was hoping to resolve this tonight, but I have run out of gas. More tomorrow.

**Comment by stgraber** [comment]
Did you get to run the last binary I gave you? That should at least tell us what command is returning that error.

**Comment by sbworth** [comment]
I didn't run it. I have discovered one container that does not have a LV as backing store, despite the fact that it was created recently. I was going to investigate a bit more. Is it possible that one of the other minor releases in the 2.8 series had a bug that could have resulted in a dir backing store instead of an LV?
Is it possible for an individual container to have a container-specific setting that differs from the default? (I certainly did not do it on purpose.)
I think that container is what stopped LXD in the last run, because it is alphabetically next after the successful sequence; so we may be very close.

**Comment by stgraber** [comment]
I can't remember any bug in the LVM code which would have caused LXC to create a directory backed container rather than LVM when the storage.lvm_vg_name key is set.

It's the kind of thing that could happen with the btrfs happen because LXD would silently fallback if btrfs isn't detected anymore, but for LVM we only check the storage key...

**Comment by brauner** [comment]
You could probably cause this by creating a directory container and then switching to LVM by setting the config key. Will try to reproduce. In any case, we might need to handle this corner case as well and revert dir-backed containers detected during upgrade to LVM-backed containers.

**Comment by brauner** [comment]
So I managed to create a mixed-storage LXD instance {DIR, LVM} by doing the following:
[![asciicast](https://asciinema.org/a/106091.png)](https://asciinema.org/a/106091)

So this means we should handle this case as well and migrate DIR-backed containers to LVM-backed containers on upgrade.

**Comment by brauner** [comment]
Actually, it might be possible to create these mixed-storage instances in any {LVM,<valid-storage-type>} combinations. Switching from LVM back should not be possible since LXD will refuse.

**Comment by brauner** [comment]
Yup, reproduced creating a zfs pool first, creating a container and then switching to lvm on a `2.8` instance.

**Comment by brauner** [comment]
I've got code coming that will handle this.

**Comment by sbworth** [comment]
@brauner I am afraid that it is even stranger than that, as the non-LVM container was clearly created long after many of the LVM containers. My best guess is that I created it on my other LXD server, which has only dir storage and migrated it to the LVM-backed server. Normally, this resulted in an LVM server on the receiving end, but in this one case I ended up with a dir server.

In preparation for the multi-storage support, was there a release in which you moved the storage setting from a global variable to a per-container variable, which might have been picked up -- and implemented -- by my LVM-backed server, but ALSO preceding the introduction of the new storage-pools area?

Since dir storage was the original, and simplest, storage format, my LVM-backed server, if it were at the exact right version when receiving the "lxc move foo sys2:foo", could have implemented the container-specific storage type by simply writing directly into `/var/lib/lxd/containers/foo/rootfs`.

Possible?

**Comment by sbworth** [comment]
Strike my last comment. Looking inside the sqlite tables, we see that the creation date timestamp was `1480368405` which converts to 2016-11-28; so it is in fact one of the oldest containers and so probably predates the LVM conversion. It is expendable; so I am going to move it aside, delete the entries from the lxd.db and try @stgraber's latest patched LXD.

**Comment by brauner** [comment]
well, I'm sending a patch that handles such upgrades

**Comment by brauner** [comment]
So ok, here's an upgrade from a pure LVM LXD pre-storage-api LXD instance (without DIR containers in the mix):
[![asciicast](https://asciinema.org/a/106133.png)](https://asciinema.org/a/106133)

And here's an upgrade from a LVM LXD pre-storage-api LXD instance which mixes LVM and DIR containers:
[![asciicast](https://asciinema.org/a/106136.png)](https://asciinema.org/a/106136)

**Comment by brauner** [comment]
Note, that moving a DIR container to a LVM container will not set `LV Thin origin name` which you can see in `lvdisplay` output. This however, shouldn't be a problem in terms of functionality.

**Comment by sbworth** [comment]
I just ran the last LXD from @stgraber. It migrated all of the containers and renamed the volumes to containers_{foo}. It eventually bombed out on migrating the images, but I think that that may again be due to residue left behind by the pre-LVM storage.
```
WARN[03-07|09:58:29] Database already contains a valid entry for the storage pool: lxd. 
WARN[03-07|09:58:29] Storage volumes database already contains an entry for the container. 
WARN[03-07|09:58:29] Storage volumes database already contains an entry for the container. 
WARN[03-07|09:58:29] Storage volumes database already contains an entry for the container. 
WARN[03-07|09:58:29] Storage volumes database already contains an entry for the container. 
WARN[03-07|09:58:30] Storage volumes database already contains an entry for the container. 
WARN[03-07|09:58:30] Storage volumes database already contains an entry for the container. 
error: Failed to run: lvrename lxd b5b03165de7c450f5f9793c8b2eb4a364fbd81124a01511f854dd379eef52abb images_b5b03165de7c450f5f9793c8b2eb4a364fbd81124a01511f854dd379eef52abb: Existing logical volume "b5b03165de7c450f5f9793c8b2eb4a364fbd81124a01511f854dd379eef52abb" not found in volume group "lxd"
```
These new image volumes were created (names truncated for presentation):
```
  LV                  VG   Attr        LSize   Pool   Origin   Data%  Meta%
  images_11fc...d172 lxd  Vwi-a-tz-- 300.00g LXDPool   1.85
  images_18e7...5533 lxd  Vwi-a-tz--  10.00g LXDPool   8.38
  images_457a...6595 lxd  Vwi-a-tz--  10.00g LXDPool   5.65
  images_543e...3d15 lxd  Vwi-a-tz--  10.00g LXDPool   8.59
  images_a570...e869 lxd  Vwi-a-tz--  10.00g LXDPool   13.85
```
and reflected in `/var/lib/lxd/storage-pools/lxd/images` as a bunch of empty directories:
```
root@sys2:/var/lib/lxd/storage-pools/lxd# ls -l images
total 24
drwx------ 2 root root 4096 Mar  7 09:58 11fc...d172
drwx------ 2 root root 4096 Mar  7 09:58 18e7...5533
drwx------ 2 root root 4096 Mar  7 09:58 457a...6595
drwx------ 2 root root 4096 Mar  7 09:58 543e...3d15
drwx------ 2 root root 4096 Mar  7 09:58 a570...3e869
drwx------ 2 root root 4096 Mar  7 09:58 b5b0...52abb
```
My `/var/lib/lxd/images` looks like it is full of cruft. (Again filenames truncated for presentation.):
```
root@sys2:/var/lib/lxd/images# ls -l
total 1448392
-rw-r--r-- 1 root root       844 Mar  3 21:37 11fc...d172
-rw-r--r-- 1 root root 153494344 Mar  3 21:40 11fc...d172.rootfs
-rw-r--r-- 1 root root 199032954 Jan 10 10:45 18e7...5533
-rw-r--r-- 1 root root       600 Jan  9 13:47 457a...6595
-rw-r--r-- 1 root root  98614308 Jan  9 13:47 457a...6595.rootfs
-rw-r--r-- 1 root root       596 Nov 24 23:36 543e...3d15
-rw-r--r-- 1 root root 130996352 Nov 24 23:36 543e...3d15.rootfs
-rw-r--r-- 1 root root 418665102 Nov 30 15:05 a570...e869
-rw-r--r-- 1 root root       816 Nov 24 15:42 b5b0...2abb
-rw-r--r-- 1 root root 125798024 Nov 24 15:42 b5b0...2abb.rootfs
-rw-r--r-- 1 root root       600 Nov 24 23:40 bfd1...f9f8
-rw-r--r-- 1 root root   2487084 Nov 24 23:40 bfd1...f9f8.rootfs
-rw-r--r-- 1 root root       596 Jan  7 17:51 d7c1...3cc6
lrwxrwxrwx 1 root root        73 Jan  7 17:51 d7c1...3cc6.lv -> /dev/lxd/d7c16c4fedd3308b5bffdb91f491b8458610c6115d37ace8ba4bcf5c29b23cc6
-rw-r--r-- 1 root root   2756944 Jan  7 17:51 d7c1...3cc6.rootfs
-rw-r--r-- 1 root root 285455980 Feb 14 13:28 e12c...eeed
lrwxrwxrwx 1 root root        73 Feb 14 13:28 e12c...eeed.lv -> /dev/lxd/e12c3c1aed259ce62b4a5e8dc5fe8b92d14d36e611b3beae3f55c94df069eeed
-rw-r--r-- 1 root root       596 Nov 24 23:39 ff52...3ac28
-rw-r--r-- 1 root root  65777664 Nov 24 23:39 ff52...3ac28.rootfs
```
I'm thinking that I may just need to move aside all of the images that have no reference to a logical volume and then run LXD again. Does that seem plausible to you guys?

Here is some formatted data from lxd.db images tables:
```
sqlite> select id,cached,fingerprint,filename,creation_date,last_use_date from images;
| id | cached | fingerprint | filename                                      | creation_date             | last_use_date                       |
|  2 |      0 | b5b0...2abb | ubuntu-14.04-server-cloudimg-amd64-lxd.tar.xz | 2016-11-09 00:00:00+00:00 |                                     |
|  3 |      0 | 543e...3d15 | lxd.tar.xz                                    | 2016-11-25 00:00:00+00:00 | 2016-11-30 21:37:31.630646328+00:00 |
|  4 |      0 | ff52...ac28 | lxd.tar.xz                                    | 2016-11-25 00:00:00+00:00 |                                     |
|  8 |      0 | a570...e869 |                                               | 0001-01-01 00:00:00+00:00 |                                     |
| 50 |      0 | d7c1...3cc6 | lxd.tar.xz                                    | 2017-01-07 00:00:00+00:00 | 2017-01-13 19:34:34.399231656+00:00 |
| 51 |      0 | 457a...6595 | lxd.tar.xz                                    | 2017-01-08 00:00:00+00:00 | 2017-01-09 18:48:43.133706794+00:00 |
| 52 |      1 | 18e7...5533 |                                               | 0001-01-01 00:00:00+00:00 | 2017-01-10 15:45:22.877654385+00:00 |
| 55 |      0 | e12c...eeed |                                               | 0001-01-01 00:00:00+00:00 | 2017-02-14 18:33:03.41803276+00:00  |
| 60 |      0 | 11fc...d172 | ubuntu-16.04-server-cloudimg-amd64-lxd.tar.xz | 2017-03-03 00:00:00+00:00 | 2017-03-04 03:47:50.051114525+00:00 |
```
Finally, the contents of lxd.db table storage_volumes. Note that the images all come at the end with the partially obfuscated containers above. I also see references to containers that were deleted (with foreign_keys=ON):
```
id|name|storage_pool_id|type
1|a.....3|1|0
2|a...4|1|0
3|a.......d|1|0
4|a.....w|1|0
5|a......t|1|0
6|b.....r|1|0
7|c.............e|1|0
8|d...|1|0
9|e......t|1|0
10|f......5|1|0
11|foo|1|0
12|h........w|1|0
13|h........w|1|0
14|j....2|1|0
15|l.........e|1|0
16|l....2|1|0
17|..0|1|0
18|..1|1|0
19|.......2|1|0
20|.......4|1|0
21|p......p|1|0
22|p........l|1|0
23|p.....t|1|0
24|p...........2|1|0
25|p..|1|0
26|s.............4|1|0
27|s...........e|1|0
28|test1|1|0
29|test2|1|0
30|u.....s|1|0
31|v...t|1|0
32|v....1|1|0
33|v...|1|0
34|w.....k|1|0
35|11fc1b1d39b9f9cd7e9491871f1421ac4278e1d599ecf5d180f2a6e2483bd172|1|1
36|18e7ed74d0d653894f65343afbc35b92c6781933c273943d882c36a5c5535533|1|1
37|457a80ea4720900b69e5542cea5351f58021331bc96e773e4855a3e2ce1e6595|1|1
38|543e662b70958f5b87f68b20eb0a205d8c4b14c41f80699e9a98b3b851883d15|1|1
39|a570ce23e1dae791e7b8b2f2bcb98c1404273e97c7a1fb972bf0f5835ac3e869|1|1
40|b5b03165de7c450f5f9793c8b2eb4a364fbd81124a01511f854dd379eef52abb|1|1
```
OK. Enough for now.

**Comment by brauner** [comment]
Do
- `/dev/lxd/d7c16c4fedd3308b5bffdb91f491b8458610c6115d37ace8ba4bcf5c29b23cc6`
- `/dev/lxd/e12c3c1aed259ce62b4a5e8dc5fe8b92d14d36e611b3beae3f55c94df069eeed`
exist? If not, you can delete them and re-run the upgrade. The image directory itself should only contain plain files of the format `<sha256>` and `<sha256>.rootf`. These should be kept around!

**Comment by brauner** [comment]
Ah, I see the upgrade failure may well be due, as you correctly observed, that the image used to create the dir container is still around and does not exist as a LVM logical volume. The new upgrade code for mixed-storage LXD instances should handle this.

**Comment by sbworth** [comment]
Yes, those images exist and are dependencies of some containers. I was just putting together an orderly table view of the output of lvs:
```
| LV                          | VG  | Attr               | LSize   | Pool                     | Origin | Data% |
|-----------------------------+-----+--------------------+---------+--------------------------+--------+-------|
| LXDPool                     | lxd | twi-aotz--   2.56t |         |                          |   3.92 |  2.13 |
| containers_......3          | lxd | Vwi-aotz--  10.00g | LXDPool |                          |  21.34 |       |
| containers_....4            | lxd | Vwi-aotz--  10.00g | LXDPool |                          |  53.18 |       |
| containers_.......w         | lxd | Vwi-aotz--  10.00g | LXDPool | images_18e7...5533 42.01 |        |       |
| containers_.......t         | lxd | Vwi-aotz--  10.00g | LXDPool |                          |  10.28 |       |
| containers_......r          | lxd | Vwi-aotz--  10.00g | LXDPool |                          |  13.78 |       |
| containers_...............e | lxd | Vwi-aotz--  10.00g | LXDPool | images_543e...3d15 8.77  |        |       |
| containers_.......t         | lxd | Vwi-aotz--  10.00g | LXDPool |                          |  16.87 |       |
| containers_.......5         | lxd | Vwi-aotz--  10.00g | LXDPool |                          |  31.73 |       |
| containers_...              | lxd | Vwi-aotz-- 300.00g | LXDPool | images_11fc...d172 1.85  |        |       |
| containers_..........w      | lxd | Vwi-aotz--  10.00g | LXDPool |                          |  19.22 |       |
| containers_..........w      | lxd | Vwi-aotz--  10.00g | LXDPool |                          |  11.45 |       |
| containers_.....2           | lxd | Vwi-aotz--  10.00g | LXDPool |                          |   9.85 |       |
| containers_...........e     | lxd | Vwi-aotz--  10.00g | LXDPool |                          |  40.40 |       |
| containers_.....2           | lxd | Vwi-aotz--  10.00g | LXDPool |                          |  98.77 |       |
| containers_..0              | lxd | Vwi-aotz-- 300.00g | LXDPool |                          |   1.94 |       |
| containers_..1              | lxd | Vwi-aotz-- 300.00g | LXDPool |                          |   1.93 |       |
| containers_.......2         | lxd | Vwi-aotz--  10.00g | LXDPool |                          |  11.84 |       |
| containers_.......4         | lxd | Vwi-aotz-- 300.00g | LXDPool |                          |   1.88 |       |
| containers_........p        | lxd | Vwi-aotz-- 300.00g | LXDPool |                          |   1.85 |       |
| containers_..........l      | lxd | Vwi-aotz-- 300.00g | LXDPool |                          |   1.90 |       |
| containers_......t          | lxd | Vwi-aotz--  10.00g | LXDPool | images_543e...3d15 8.59  |        |       |
| containers_.............2   | lxd | Vwi-aotz-- 300.00g | LXDPool |                          |   1.95 |       |
| containers_..b              | lxd | Vwi-aotz-- 300.00g | LXDPool |                          |   2.11 |       |
| containers_...............4 | lxd | Vwi-aotz--  10.00g | LXDPool | d7c1...3cc6        3.42  |        |       |
| containers_.............e   | lxd | Vwi-aotz--  10.00g | LXDPool | images_457a...6595 5.71  |        |       |
| containers_....1            | lxd | Vwi-aotz--  10.00g | LXDPool |                          |  12.15 |       |
| containers_....2            | lxd | Vwi-aotz--  10.00g | LXDPool |                          |   9.62 |       |
| containers_......s          | lxd | Vwi-aotz--  10.00g | LXDPool |                          |  69.05 |       |
| containers_....t            | lxd | Vwi-aotz--  10.00g | LXDPool |                          |  12.34 |       |
| containers_.....1           | lxd | Vwi-aotz-- 300.00g | LXDPool |                          |   1.85 |       |
| containers_...1             | lxd | Vwi-aotz-- 300.00g | LXDPool |                          |   1.88 |       |
| containers_......k          | lxd | Vwi-aotz--  10.00g | LXDPool |                          |  42.80 |       |
| d7c1...3cc6                 | lxd | Vwi-a-tz--  10.00g | LXDPool |                          |   2.95 |       |
| e12c...eeed                 | lxd | Vwi-a-tz-- 300.00g | LXDPool |                          |   1.89 |       |
| f..........1                | lxd | Vwi-aotz--  40.00g | LXDPool |                          |  34.53 |       |
| images_11fc...d172          | lxd | Vwi-a-tz-- 300.00g | LXDPool |                          |   1.85 |       |
| images_18e7...5533          | lxd | Vwi-a-tz--  10.00g | LXDPool |                          |   8.38 |       |
| images_457a...6595          | lxd | Vwi-a-tz--  10.00g | LXDPool |                          |   5.65 |       |
| images_543e...3d15          | lxd | Vwi-a-tz--  10.00g | LXDPool |                          |   8.59 |       |
| images_a570...e869          | lxd | Vwi-a-tz--  10.00g | LXDPool |                          |  13.85 |       |
| lphys-home                  | lxd | Vwi-aotz-- 400.00g | LXDPool |                          |   2.82 |       |
```

**Comment by brauner** [comment]
Cool, yeah those other image failures are caused by them not being present as LVM logical volumes since they were used to create dir containers. Thanks for your help!

**Comment by sbworth** [comment]
I'm glad that my pain could be of value. ;-)

**Comment by sbworth** [comment]
If I see:
```
-rw-r--r-- 1 root root     596 Jan  7 17:51 d7c16c4fedd3308b5bffdb91f491b8458610c6115d37ace8ba4bcf5c29b23cc6
lrwxrwxrwx 1 root root      73 Jan  7 17:51 d7c16c4fedd3308b5bffdb91f491b8458610c6115d37ace8ba4bcf5c29b23cc6.lv -> /dev/lxd/d7c16c4fedd3308b5bffdb91f491b8458610c6115d37ace8ba4bcf5c29b23cc6
-rw-r--r-- 1 root root 2756944 Jan  7 17:51 d7c16c4fedd3308b5bffdb91f491b8458610c6115d37ace8ba4bcf5c29b23cc6.rootfs
```
and absent any dir-backed containers, are the plain file versions just leftovers that we can safely clear away. (Perhaps LXD should do the clearing to ensure database consistency, but I am thinking more narrowly of not missing any container dependencies.)

**Comment by brauner** [comment]
That depends on whether the corresponding image has an entry in the images database. If that is the case then simply deleting the image from the folder will leave you with an entry for this image in the db. As soon as @stgraber is around he can provide you with a pre-built binary based on my patches that will likely take care of this issue. If you can't wait that long, then do:
- check if the image you are trying to remove has a corresponding logical volume
 - yes --> leave it
 - no --> check if image has an entry in the images database and in the storage volumes database if it has delete the image entries in `/var/lib/lxd/images/<sha256>{.rootfs}` and the corresponding entries in the db

**Comment by sbworth** [comment]
I am content to wait for now for the @stgraber updated LXD. I simply want to understand the dependencies as well as possible for future reference. I'm going to create a new issue with a suggestion for tweaking your handling of image/container updates with safety in mind. (You guys, of course, get to decide whether it make sense.)

**Comment by stgraber** [comment]
Sorry about that, worked till 3am so woke up pretty late today :)
Will build you a new binary now.

**Comment by stgraber** [comment]
@sbworth 

Binary updated:
https://dl.stgraber.org/lxd-3026
sha256: ad96813bb5ecc29dde483b48ee284682df3f128d7b1006f2c313300585970bdf

**Comment by sbworth** [comment]
@stgraber We have success.

Should I continue to run this lxd, or should the 2.10.1 version now suffice until the next upgrade?

**Comment by stgraber** [comment]
SInce the changes we made were limited to the upgrade code and you've now gotten past that upgrade, you can resume using 2.10.1. We'll have LXD 2.11 out later today with the fix.

**Comment by stgraber** [comment]
And merged that last batch of fixes from @brauner so all the fixes needed to sort out this issue are now in the master branch, closing this issue.

**Comment by sbworth** [comment]
Now running successfully with 2.10.1 lxd in post-migration state.

Thanks again @stgraber @brauner .
